### PR TITLE
Test and improve Xapi periodic scheduler

### PR DESCRIPTION
**Pull request by freddy77**
There was some issue in the current code, from structure corruption to thread safety.
Add some test and fix discovered issues.

More details on commit messages.

### CODE REVIEW

File: ocaml/libs/xapi-stdext/lib/xapi-stdext-threads/scheduler.ml
Comment (psafont) [inline ocaml/libs/xapi-stdext/lib/xapi-stdext-threads/scheduler.ml:96]: Didn't this use to be 10 seconds only?

Comment (freddy77) [reply ocaml/libs/xapi-stdext/lib/xapi-stdext-threads/scheduler.ml:96]: Yes, forgot to put it back although the idea should be to not polling and having a more stable Delay implementation to avoid the fallback (I was thinking to put something like one day then).

But for the moment I'll be change back to 10 seconds.

File: ocaml/libs/xapi-stdext/lib/xapi-stdext-threads/scheduler.ml
Comment (psafont) [inline ocaml/libs/xapi-stdext/lib/xapi-stdext-threads/scheduler.ml:85]: I don't think this needs nested matching:
```suggestion
  | Some ({ty=Periodic timer; _} as ev) ->
      let ( ++ ) = Clock.add_span in
      let item = {Ipq.ev; Ipq.time= Mtime_clock.now () ++ timer} in
      Ipq.add queue item
      pending_event := None
  | Some {ty=OneShot; _} ->
      pending_event := None
  | None ->
      ()
```

**Review by psafont — APPROVED** [review]

**Comment by edwintorok** [comment]
There are some changes in common here with my changes (https://github.com/xapi-project/xen-api/compare/feature/perf...edwintorok:xen-api:pr/mtime), for which I was about to open a PR (and haven't opened one before because there were too many PRs open).

I think we should probably open all PRs that we have, in a draft form so that we're aware of each-other's work. Otherwise we're just duplicating work. We have the Mtime changes done by 3 different people so far (@psafont , you and myself).

**Review by edwintorok — COMMENTED** [review]
'Do not wait huge amount of time if the queue is empty' was already fixed by https://github.com/xapi-project/xen-api/pull/6122/files#diff-5c3ab2a7766ba1f2c349d7985a633259d244dd8d5b2a3dfaf948337dc9a45221 (although that is on `feature/perf`, so likely the 2 changes here will conflict.)

File: ocaml/libs/xapi-stdext/lib/xapi-stdext-threads/scheduler.ml
Comment (edwintorok) [inline ocaml/libs/xapi-stdext/lib/xapi-stdext-threads/scheduler.ml:90]: Yes, looking at the maximum and popping really needs to be done while still holding the lock, otherwise there was a race here previously.

**Review by edwintorok — COMMENTED** [review]
Would be useful to split the bugfixes into separate commits (like I've done in https://github.com/xapi-project/xen-api/pull/6161/commits), that way it'll be easier to see which changes are duplicate (and one of the commits can be dropped from one of the PRs, and which ones are still needed).

File: ocaml/libs/xapi-stdext/lib/xapi-stdext-threads/scheduler.ml
Comment (edwintorok) [inline ocaml/libs/xapi-stdext/lib/xapi-stdext-threads/scheduler.ml:103]: What if there already was a pending_event? This would overwrite it.

Comment (edwintorok) [reply ocaml/libs/xapi-stdext/lib/xapi-stdext-threads/scheduler.ml:103]: Probably best to avoid introducing more mutation, and pass this value directly to add_periodic_pending instead of relying on a global for parameter passing.

File: ocaml/libs/xapi-stdext/lib/xapi-stdext-threads/scheduler.ml
Comment (edwintorok) [inline ocaml/libs/xapi-stdext/lib/xapi-stdext-threads/scheduler.ml:76]: This will slowly drift over time if XAPI is busy, or handling the periodic event takes a while.
OTOH if we would strictly trigger handling of this event at a constant rate then we might end up busy looping (if we don't finish handling the event by the time it is scheduled again).

To get back on track we can try finding the next point in time where `initial_delay + N*interval = time`. 

This can be done in a separate/commit PR though.

File: ocaml/libs/xapi-stdext/lib/xapi-stdext-threads/scheduler.ml
Comment (freddy77) [reply ocaml/libs/xapi-stdext/lib/xapi-stdext-threads/scheduler.ml:103]: No, it's not overridden, there's only one pending event.

No, it can't be stored and passed to `add_periodic_pending` otherwise you could not remove while processing.

File: ocaml/libs/xapi-stdext/lib/xapi-stdext-threads/scheduler.ml
Comment (freddy77) [reply ocaml/libs/xapi-stdext/lib/xapi-stdext-threads/scheduler.ml:76]: This was done intentionally, it's the same behaviour as it was before. I though it was by design.

**Comment by edwintorok** [comment]
The ones that are difficult to split you can leave as is. I'll rebase my changes on top of this one, once it is merged to master.

**Comment by freddy77** [comment]
Rebased on master and split the multiple fix patch.

**Comment by freddy77** [comment]
Rebased on `master`, added a new test.

File: ocaml/libs/xapi-stdext/lib/xapi-stdext-threads/scheduler.ml
Comment (edwintorok) [reply ocaml/libs/xapi-stdext/lib/xapi-stdext-threads/scheduler.ml:103]: Why don't we call the body of add_periodic_pending while holding the lock? That'd avoid the race condition that was there before with remove_queue, and you wouldn't have to introduce a new global variable.

**Review by edwintorok — APPROVED** [review]
Lets leave it the global `pending_event` as it is for now, and I'll revisit this when I introduce a Picos-based scheduler in XAPI.
Picos uses a more efficient [psq](https://ocaml.org/p/psq/latest/doc/Psq/index.html) based implementation, rather than the array based implementation in XAPI's Ipq (which afaict needs to be grown and resized every time we add and remove elements).

And we can probably implement 'removing from the queue' of a periodic task as canceling the Fiber that is running the periodic task (which internally uses scheduler delays to reschedule itself).

**Comment by freddy77** [comment]
> Lets leave it the global `pending_event` as it is for now, and I'll revisit this when I introduce a Picos-based scheduler in XAPI. Picos uses a more efficient [psq](https://ocaml.org/p/psq/latest/doc/Psq/index.html) based implementation, rather than the array based implementation in XAPI's Ipq (which afaict needs to be grown and resized every time we add and remove elements).

No, it does not need to resize every time. But the lookup is `O(n)` instead of `O(log(n))`.

Yes, calling `add_periodic_pending` is not done while extracting the minimal to avoid regressions.

**Comment by edwintorok** [comment]
I thought I pressed the merge button, but maybe it didn't work for some reason. This means I'll need to rebase my PR again because it probably didn't include this change yet https://github.com/xapi-project/xen-api/pull/6161

**Pull request merged**
